### PR TITLE
[Core] Allow usage of ChibiOS's SIO driver for split keyboards

### PR DESCRIPTION
**Pull request by KarlK90**
This PR build upon https://github.com/qmk/qmk_firmware/pull/16647 while this one is not merged please ignore the commit [Use a mutex for split shared memory synchronization](https://github.com/qmk/qmk_firmware/pull/15907/commits/0829b51c74fbd6839b28fec1c6c0c2027730bc2e) in this PR.

## Description

Allow usage of 21.11+ ChibiOS's SIO driver for split keyboards. The cool thing about the SERIAL and SIO driver is that both implement the [`BaseChannel`](https://chibiforge.org/doc/21.11/hal/group___i_o___c_h_a_n_n_e_l.html) interface that provides blocking reading and writing methods that resemble what we already use in the driver. Only some low level aspects like initialization have to be handled differently.

### Supported Hardware (incomplete)

**Every STM32 MCU family that supports `USARTv2` or `USARTv3` like:**

* STM32F0xx (Patches needed to enable SIO IRQ handler) 
* STM32F3xx (Patches needed to enable SIO IRQ handler)
* STM32G0xx
* STM32G4xx 
* RP2040

### For test-driving these changes do the following:

1. Add these configs to your keyboard:

**config.h**

```c
#define SERIAL_USART_DRIVER SIOD3 // Or whatever your usart peripheral is
```

**halconf.h**
```c
#define HAL_USE_SIO TRUE
```

**mcuconf.h**
```c
#define STM32_IRQ_USART3_PRIORITY 12
#define STM32_SIO_USE_USART3 TRUE // Or what ever you usart peripheral is
```

Target use-case is the RP2040 which only implements this driver at this point. 

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* Split support for RP2040 #14877

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by KarlK90** [comment]
Updated the docs, this is now ready to be reviewed.

**Comment by KarlK90** [comment]
I re-based this PR on https://github.com/qmk/qmk_firmware/pull/16647 in order to build another upon this PR. They all depend on each other but the intent is to review them in separation.

File: docs/serial_driver.md [hunk 1]
```diff
@@ -1,129 +1,256 @@
...
-#define SOFT_SERIAL_PIN B6         // USART TX pin
-//#define USART1_REMAP             // Remap USART TX and RX pins on STM32F103 MCUs, see table below.
-#define SELECT_SOFT_SERIAL_SPEED 1 // or 0, 2, 3, 4, 5
-                                   //  0: about 460800 baud
-                                   //  1: about 230400 baud (default)
-                                   //  2: about 115200 baud
-                                   //  3: about 57600 baud
-                                   //  4: about 38400 baud
-                                   //  5: about 19200 baud
-#define SERIAL_USART_DRIVER SD1    // USART driver of TX pin. default: SD1
```
Comment (daskygit) [inline docs/serial_driver.md:64]: There's no longer any mention of setting this define, it should also be made clear that SIO would be `#define SERIAL_USART_DRIVER SIOD2` vs `#define SERIAL_USART_DRIVER SD2` for serial.

File: docs/serial_driver.md [hunk 2]
```diff
@@ -1,129 +1,256 @@
...
+```
+  LEFT                      RIGHT
++-------+      SERIAL     +-------+
+|   SSP |-----------------| SSP   |
+|       |       VDD       |       |
+|       |-----------------|       |
+|       |       GND       |       |
+|       |-----------------|       |
++-------+                 +-------+
+```
```
Comment (daskygit) [inline docs/serial_driver.md:32]: I think these diagrams may not be very clear to new users, maybe a picture using a popular controller would be a better option?

**Review by daskygit — COMMENTED** [review]
I've tested the SIO driver with my STM32G483 and it works well.

File: docs/serial_driver.md [hunk 3]
```diff
@@ -1,129 +1,256 @@
...
+## Advanced Configuration
+
+There are several advanced configuration options that can be defined in your keyboards `config.h` file:
+
+### Baudrate
+
+If you're having issues or need a higher baudrate with serial communication, you can change the baudrate which in turn controls the communication speed for serial. You want to lower the baudrate if you experience failed transactions. 
+
+```c
+#define SELECT_SOFT_SERIAL_SPEED {#}`
```
Comment (tzarc) [inline docs/serial_driver.md:221]:
```suggestion
#define SELECT_SOFT_SERIAL_SPEED {#}
```

**Review by tzarc — CHANGES_REQUESTED** [review]

File: docs/serial_driver.md (hunk 2 above)
Comment (KarlK90) [reply docs/serial_driver.md:32]: Yes indeed, although I would like to tackle that in a separate PR if that is fine with youß

File: docs/serial_driver.md (hunk 1 above)
Comment (daskygit) [reply docs/serial_driver.md:64]: Perfect ta.

**Review by tzarc — APPROVED** [review]

**Comment by KarlK90** [comment]
@daskygit as you use G4s on a crkbd could you help verifying that the `SIO` driver works as expected in Half-duplex? :slightly_smiling_face:

**Review by daskygit — APPROVED** [review]
After some [troubleshooting](https://discord.com/channels/440868230475677696/867530715192885269/987442265280626688) on Discord with Karl it's working well.

**Pull request merged**